File: api/python/examples/pdb_inspect.py
#!/usr/bin/env python
"""Inspect a Microsoft PDB (Program Database) file.

Loads a ``.pdb`` file and walks its contents: GUID/age metadata,
public symbols, user-defined types (classes) and per-compilation-unit
information (source files and functions).

Note: only available with the extended version of LIEF.

Example:

    $ python pdb_inspect.py ./program.pdb
"""

import argparse
import sys

import lief


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("file", help="Path to the PDB file")
    args = parser.parse_args()

    pdb = lief.pdb.load(args.file)
    if pdb is None:
        print(f"Error: failed to load '{args.file}'", file=sys.stderr)
        return 1

    print(f"age={pdb.age}, guid={pdb.guid}")

    for sym in pdb.public_symbols:
        print(f"name={sym.name}, section={sym.section_name}, RVA={sym.RVA}")

    for ty in pdb.types:
        if isinstance(ty, lief.pdb.types.Class):
            print(f"Class[name]={ty.name}")

    for cu in pdb.compilation_units:
        print(f"module={cu.module_name}")
        for src in cu.sources:
            print(f"  - {src}")
        for func in cu.functions:
            print(
                f"name={func.name}, section={func.section_name}, "
                f"RVA={func.RVA}, code_size={func.code_size}"
            )

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: api/rust/examples/src/bin/pdb_inspect.rs
/// This example shows how to inspect a PDB file using LIEF's API

use std::process::{self, ExitCode};
use lief::pdb::types::classlike::ClassLike;

fn main() -> ExitCode {
    if !lief::is_extended() {
        println!("This example requires the extended version of LIEF")
    }
    let mut args = std::env::args();
    if args.len() != 2 {
        println!("Usage: {} <binary>", args.next().unwrap());
        return ExitCode::FAILURE;
    }

    let path = std::env::args().last().unwrap();
    let pdb = lief::pdb::load(&path).unwrap_or_else(|| {
        process::exit(1);
    });
    println!("age={}, guid={}", pdb.age(), pdb.guid());

    for symbol in pdb.public_symbols() {
        println!("name={}, section={}, RVA={}",
            symbol.name(), symbol.section_name().unwrap_or("".to_string()),
            symbol.rva());
    }

    for ty in pdb.types() {
        if let lief::pdb::Type::Class(clazz) = ty {
            println!("Class[name]={}", clazz.name());
        }
    }

    for cu in pdb.compilation_units() {
        println!("module={}", cu.module_name());
        for src in cu.sources() {
            println!("  - {}", src);
        }

        for func in cu.functions() {
            println!("name={}, section={}, RVA={}, code_size={}",
                func.name(), func.section_name(), func.rva(), func.code_size()
            );
        }
    }


    println!("Can't process {}", path);
    ExitCode::FAILURE
}
